matron/src/device/device_list.c

```diff
@@ -72,35 +75,86 @@ void dev_list_add(device_t type, const char *path, const char *name) {
     dq.size++;
 
     union event_data *ev;
+    ev = event_data_new(event_type);
+    return ev;
+}
+
+void dev_list_add(device_t type, const char *path, const char *name) {
+    if (type < 0) {
+        return;
+    }
+
+    union event_data *ev;
+    union dev *d;
+    unsigned int midi_port_count = 0;
+
     switch (type) {
+    case DEV_TYPE_MIDI:
+        midi_port_count = dev_port_count(path);
+        for (unsigned int pidx = 0; pidx < midi_port_count; pidx++) {
+            d = dev_new(type, path, name, midi_port_count > 1, pidx);
+            ev = post_add_event(d, EVENT_MIDI_ADD);
+            if (ev != NULL) {
+                ev->midi_add.dev = d;
+                event_post(ev);
+            }
+        }
+        return;
     case DEV_TYPE_MONOME:
-        ev = event_data_new(EVENT_MONOME_ADD);
-        ev->monome_add.dev = d;
+        d = dev_new(type, path, name, true, 0);
+        ev = post_add_event(d, EVENT_MONOME_ADD);
         break;
     case DEV_TYPE_HID:
-        ev = event_data_new(EVENT_HID_ADD);
-        ev->hid_add.dev = d;
-        break;
-    case DEV_TYPE_MIDI:
-        ev = event_data_new(EVENT_MIDI_ADD);
-        ev->midi_add.dev = d;
+        d = dev_new(type, path, name, true, 0);
+        ev = post_add_event(d, EVENT_HID_ADD);
         break;
     case DEV_TYPE_CROW:
-        ev = event_data_new(EVENT_CROW_ADD);
-        ev->crow_add.dev = d;
+        d = dev_new(type, path, name, true, 0);
+        ev = post_add_event(d, EVENT_CROW_ADD);
         break;
     default:
         fprintf(stderr, "dev_list_add(): error posting event (unknown type)\n");
         return;
     }
-    event_post(ev);
+    if (ev != NULL) {
+        ev->monome_add.dev = d;
```

Comment:
seems this is using the `monome_add` structure for HID and CROW

Reply:
not sure it's relevant, but want to mention this comment from when HID was getting fixed up:
https://github.com/monome/norns/pull/662#issuecomment-447229707
Although... I can't remember specifics of the debug I did at that point.